### DISCUSSION TITLE
Configs to load cmdlets from filesystem

### DISCUSSION THREAD

**Discussion by neur1n** [wezterm/wezterm#4116] [Show and tell]
Initially, I was trying to do something like launching a predefined layout and runs `ssh ...` like tmux or zellij does. However, I think it is not safe to expose these information, such as IP addresses and passwords, in dotfiles. It seems that loading `ssh ...` and `scp ...` cmdlets from local filesystem would be safer and this requires a programmatic configuration system, so I decide to do it in wezterm which uses lua instead of (such as) kdl in zellij.

As I'm working on this, I found that I could turn it into a general cmdlet loading gadget. Basically, I bind `Ctrl+Shift+e` to bring up a prompt input for inserting the file name and split counts of the new pane.

Here is an example of spawning a new tab with one or more panes, and `ssh` into remote machines, and launch nushel, and load a nushell module. This chain is defined in a local file named *ssh127.event*.

https://github.com/wez/wezterm/assets/17579247/d7d69d09-0065-4851-9483-230c6b921031

The configuration is [here](https://github.com/Neur1n/dotfiles/blob/master/wezterm/.wezterm/n_event.lua). It's quite naive but it some how saves me some time when connecting to remote machines.

Any advices are welcome!

**Comment by wez** [comment]
Neat! Some things to consider:

* In your [load_event](https://github.com/Neur1n/dotfiles/blob/5f34929b32d53f373cfbaaf5159cb60bd508c8ac/wezterm/.wezterm/n_event.lua#L8) function, you could use [wezterm.json_parse](https://wezfurlong.org/wezterm/config/lua/wezterm/json_parse.html) to avoid having to manually parse the content of your `.event` files.  You could then do things like define a pretty name property in the .event json file and feed that into the next bit:
* Since you have a directory containing the list of possible events, consider using [wezterm.glob](https://wezfurlong.org/wezterm/config/lua/wezterm/glob.html) to get a list of the possible events (`wezterm.glob(string.format("%s/.wezterm/*.event", Wezterm.config_dir))`), and then using [InputSelector](https://wezfurlong.org/wezterm/config/lua/keyassignment/InputSelector.html) to provide a fuzzy-matchable menu list of events, instead of using `PromptInputLine`.  I know that you're passing in an optional split count with the `PromptInputLine` approach; you could pre-generate entries with reasonable split counts to do something similar with the `InputSelector`.

**Comment by neur1n** [comment]
- Great to know there is `westerm.json_parse`! I will refactor with it definitely.
- I did tried to use `wezterm.glob`, but it raised an error as that in #2401 since I don’t want to put many codes into *westerm.lua* for now.

Thanks for your hints!

**Comment by neur1n** [comment]
Changes are made in this gadget. Events are not defined in separate files anymore. I put every thing into a *snippet_vault.json* file and use `wezterm.json_parse` and `InputSelector` to provide a list of available snippets. I think it is similar to the snippet feature of Termius, but when defining a cmdlet sequence we still need to call `wezterm.sleep_ms` to make sure a cmdlet is done.

New code is [here](https://github.com/Neur1n/dotfiles/blob/59e5d86b01771c0c5ffad102c9758fcda2b6bc2c/wezterm/.wezterm/n_snippet.lua).

https://github.com/wez/wezterm/assets/17579247/7684edb5-de51-4a66-941c-7043975bfd7e
